backend-rust/src/indexer.rs

```diff
@@ -4221,19 +4225,213 @@ impl PreparedSpecialTransactionOutcomes {
     }
 }
 
+/// The `SpecialEvents` of a payday block in the order they
+/// occur in the block.
+struct PreparedPaydaySpecialTransacionOutcomes {
+    /// Height of the payday block containing the events.
+    block_height: i64,
+    events:       Vec<SpecialTransactionOutcome>,
+}
+
+impl PreparedPaydaySpecialTransacionOutcomes {
+    fn prepare(block_height: i64, events: &[SpecialTransactionOutcome]) -> anyhow::Result<Self> {
+        Ok(Self {
+            block_height,
+            events: events.to_vec(),
+        })
+    }
+
+    async fn save(
+        &self,
+        tx: &mut sqlx::Transaction<'static, sqlx::Postgres>,
+    ) -> anyhow::Result<()> {
+        // Extract the rewards from the `SpecialEvents` in each payday block
+        // and associate it with the `pool_owner`.
+        // The `pool_owner` can be either a `baker_id` or `None`.
+        // The `pool_owner` is `None` if the pool rewards are for the passive
+        // delegators which can happen at most once per payday block.
+        //
+        // https://docs.rs/concordium-rust-sdk/6.0.0/concordium_rust_sdk/types/enum.SpecialTransactionOutcome.html#variant.PaydayAccountReward
+        // The order of `SpecialEvents` in each payday block has a meaning to
+        // determine which rewards go to the baker of a baker pool and which
+        // rewards go to the pool's delegators.
+        //
+        // For example:
+        // PaydayPoolReward to pool 1
+        // PaydayAccountReward to account 5
+        // PaydayAccountReward to account 6
+        // PaydayAccountReward to account 1
+        // PaydayPoolReward to pool 8
+        // PaydayAccountReward to account 8
+        // PaydayAccountReward to account 2
+        // PaydayPoolReward to `None`
+        // PaydayAccountReward to account 10
+        // PaydayAccountReward to account 3
+        // Means 5, 6 are receiving rewards from delegating to 1 and 2 receiving rewards
+        // from delegating to 8, and 10, 3 are receiving rewards from passive
+        // delegation.
+        //
+        // `PaydayPoolReward` and `PaydayAccountReward` events only occure in payday
+        // blocks starting in protocol 4.
+        //
+        // The mapping `acc_payday_pool_rewards` associates the `pool_owner` (a
+        // `baker_id` or `None`) to its `PaydayPoolRewards`. The `None` value
+        // tracks passive deleagation.
+        let mut acc_payday_pool_rewards: HashMap<Option<u64>, PaydayPoolRewards> = HashMap::new();
```

Comment:
Instead of building this hashmap and then traversing it to build the vecs, we could build the vecs directly since we rely on the order of the events anyway.

Reply:
I optimized the indexer performance with regards to the payday rewards insertion.